lib/trie/trie_test.go

```diff
@@ -1,1479 +1,3081 @@
-
-	// Only the current trie should have a different root hash since it is updated.
-	require.NotEqual(t, ssTrie, dcTrieHash)
-	require.NotEqual(t, ssTrie, tHash)
-	require.Equal(t, dcTrieHash, tHash)
-}
+			deleted, allDeleted := trie.ClearPrefixLimit(testCase.prefix, testCase.limit)
 
-func TestTrie_ClearPrefixVsDelete(t *testing.T) {
-	prefixes := [][]byte{
-		{},
-		{0x0},
-		{0x01},
-		{0x01, 0x35},
-		{0xf},
-		{0xf2},
-		{0x01, 0x30},
-		{0x01, 0x35, 0x70},
-		{0x01, 0x35, 0x77},
-		{0xf2, 0x0},
-		{0x07},
-		{0x09},
-		[]byte("a"),
+			assert.Equal(t, testCase.deleted, deleted)
+			assert.Equal(t, testCase.allDeleted, allDeleted)
+			assert.Equal(t, testCase.expectedTrie, trie)
+		})
 	}
+}
 
-	cases := [][]Test{
-		{
-			{key: []byte{0x01, 0x35}, value: []byte("pen")},
-			{key: []byte{0x01, 0x35, 0x79}, value: []byte("penguin")},
-			{key: []byte{0x01, 0x35, 0x7}, value: []byte("g")},
-			{key: []byte{0x01, 0x35, 0x99}, value: []byte("h")},
-			{key: []byte{0xf2}, value: []byte("feather")},
-			{key: []byte{0xf2, 0x3}, value: []byte("f")},
-			{key: []byte{0x09, 0xd3}, value: []byte("noot")},
-			{key: []byte{0x07}, value: []byte("ramen")},
-			{key: []byte{0}, value: nil},
-		},
-		{
-			{key: []byte{0x01, 0x35}, value: []byte("pen")},
-			{key: []byte{0x01, 0x35, 0x79}, value: []byte("penguin")},
-			{key: []byte{0x01, 0x35, 0x70}, value: []byte("g")},
-			{key: []byte{0xf2}, value: []byte("feather")},
-			{key: []byte{0xf2, 0x30}, value: []byte("f")},
-			{key: []byte{0x09, 0xd3}, value: []byte("noot")},
-			{key: []byte{0x07}, value: []byte("ramen")},
-		},
-		{
-			{key: []byte("asdf"), value: []byte("asdf")},
-			{key: []byte("ghjk"), value: []byte("ghjk")},
-			{key: []byte("qwerty"), value: []byte("qwerty")},
-			{key: []byte("uiopl"), value: []byte("uiopl")},
-			{key: []byte("zxcv"), value: []byte("zxcv")},
-			{key: []byte("bnm"), value: []byte("bnm")},
+func Test_Trie_clearPrefixLimit(t *testing.T) {
+	t.Parallel()
+
+	testCases := map[string]struct {
+		trie          Trie
+		parent        Node
+		prefix        []byte
+		limit         uint32
+		expectedLimit uint32
+		newParent     Node
+		updated       bool
+		allDeleted    bool
+		expectedTrie  Trie
+	}{
+		"limit is zero": {
+			allDeleted: true,
+		},
+		"nil parent": {
+			limit:         1,
+			expectedLimit: 1,
+			allDeleted:    true,
+		},
+		"leaf parent with common prefix": {
+			parent: &node.Leaf{
+				Key: []byte{1, 2},
+			},
+			prefix:     []byte{1},
+			limit:      1,
+			updated:    true,
+			allDeleted: true,
+		},
+		"leaf parent with key equal prefix": {
+			parent: &node.Leaf{
+				Key: []byte{1},
+			},
+			prefix:     []byte{1},
+			limit:      1,
+			updated:    true,
+			allDeleted: true,
+		},
+		"leaf parent with key no common prefix": {
+			parent: &node.Leaf{
+				Key: []byte{1, 2},
+			},
+			prefix:        []byte{1, 3},
+			limit:         1,
+			expectedLimit: 1,
+			newParent: &node.Leaf{
+				Key: []byte{1, 2},
+			},
+			allDeleted: true,
```

Comment:
why is `allDeleted` true here? it looks like nothing is deleted? same with below few cases

Reply:
It is all deleted since we want to delete at prefix `[]byte{1, 3}` but there is nothing there already 😉 It's probably the same for the other cases?

Reply:
hmm interesting, we might want to make sure that behaviour is the same as substrate